feat: Implement Multiprovider feature

Tmakinde opened the pull request:
## Summary

Adds a new feature that lets you plug in more than one feature flag provider at the same time. Instead of picking one provider up front, you can give the SDK a list, and it will try them in a defined order (or pattern) and return the first good answer. If none succeed, you get a single combined error.

## Why this is useful

Real apps often use multiple sources for flags (primary remote service, fallback local file, experimentation provider, etc.). This feature avoids hand-written chaining logic and gives a simple, uniform way to try multiple providers.

## How it works (plain flow)

1. You create a `Multiprovider` and pass an array of providers. You may optionally name each provider. If you do not name one, the code gives it a unique name automatically.
2. You call `resolveXValue(...)` (boolean, string, int, float, object) on the multiprovider.
3. A strategy decides which providers to try and when to stop.
4. Each provider is asked for the flag value (one at a time for the default strategy).
5. If one returns a successful value, that result is returned immediately.
6. If all fail (throw or error), a single error result is returned summarizing the failure.

## Main classes in simple terms

- `Multiprovider`: The coordinator. Holds all the providers, calls them in order, chooses a strategy, and returns the final answer.
- `BaseEvaluationStrategy`: A template for “how to walk through the list.” It defines the decision points (should we try this provider? should we move to the next? how do we pick the final result?).
- `FirstMatchStrategy` (default): Tries providers in order, returns the first success, stops early.
- `FirstSuccessfulStrategy`: Alternative strategy with slightly different stop rules (still about first success).
- `ProviderResolutionResult`: A small wrapper for one provider’s attempt (either a value or an error).
- `FinalResult`: The final outcome after all attempts (either a single good value or a bundle of errors).
- `StrategyEvaluationContext` / `StrategyPerProviderContext`: Helper objects carrying flag info and provider-specific context so strategies can make simple yes/no decisions.

## Backward compatibility

- Existing code using a single provider is unchanged.
- No existing public APIs were removed or altered.
- The new classes are additive and live in their own folder.
- If you do nothing, the SDK works exactly as before.

## Testing

What was added:

- Tests for the multiprovider basic resolution path.
- Tests for strategies (ensuring the first matching provider wins; error path when all fail).
- Tests for final result object (success vs aggregated errors).
- Tests for validation (rejects duplicate explicit names, rejects empty name strings).

## Simpler example

```php
$multi = new Multiprovider([
    ['provider' => $primary],
    ['name' => 'audit', 'provider' => $audit],
    ['provider' => $fallback],
]);

$result = $multi->resolveBooleanValue('beta-feature', false);

$value = $result->getValue(); // either a resolved boolean or the default
$error = $result->getError(); // null if success
```

## Follow-up improvements (non-blocking)

- README section explaining multi-provider setup.

Tmakinde committed:
feat: Created methods to validate and register providers input

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: Created methods to validate and register providers input

Signed-off-by: tmakinde <makindet74@gmail.com>

tcarrio commented:
Hey @Tmakinde , initial thoughts here is that other languages utilized the contrib repository for implementing the multi provider. The reasoning as I understand it is it's an extension of the primary provider, not a function of it itself.

That said it would be fairly straightforward to add support for it [there](https://GitHub.com/open-feature/php-sdk-contrib)

@beeme1mr is that the expectation for all languages?

Tmakinde commented:
> Hey @Tmakinde , initial thoughts here is that other languages utilized the contrib repository for implementing the multi provider. The reasoning as I understand it is it's an extension of the primary provider, not a function of it itself.
> 
> That said it would be fairly straightforward to add support for it [there](https://GitHub.com/open-feature/php-sdk-contrib)
> 
> @beeme1mr is that the expectation for all languages?

Ooh.

Please let me know the best place to implement this.
@beeme1mr told me it should be in `php-sdk` repo.

Please confirm.

Tmakinde commented:
> We've been slowly moving the multi-providers to the SDKs themselves. The logic being that it's easier to access and manage provider state from within the SDK and it shouldn't add too much complexity to the SDKs. The [.NET SDK](https://github.com/open-feature/dotnet-sdk#multi-provider) is a good example of what we're looking for.

Thank you @beeme1mr

Tmakinde committed:
feat: Add multiprovider strategy files

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: Add multiprovider strategy files

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: add multiprovider final result implementation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: add multiprovider final result implementation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: add a single provider resolution implementation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: add a single provider resolution implementation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: resolve providers through strategy and return proper error as stated in documentation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
feat: resolve providers through strategy and return proper error as stated in documentation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
fix: Refactor strategy implementation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
fix: Refactor strategy implementation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: refactor provider validation method

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: refactor provider validation method

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
test: Add test for provider and it strategy

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
test: Add test for provider and it strategy

Signed-off-by: tmakinde <makindet74@gmail.com>

tcarrio commented on src/implementation/multiprovider/Multiprovider.php:
```diff
@@ -0,0 +1,359 @@
+
+            $details = match ($flagType) {
+                'boolean' => is_bool($defaultValue)
+                ? $provider->resolveBooleanValue($context->getFlagKey(), $defaultValue, $evalContext)
```
These default value checks are unnecessarily redundant. You could validate default values at the start if they're going to throw.

tcarrio replied:
That also drastically improves the match block readability.

tcarrio commented on src/implementation/multiprovider/strategy/BaseEvaluationStrategy.php:
```diff
@@ -0,0 +1,79 @@
+ */
+abstract class BaseEvaluationStrategy
+{
+    public string $runMode = 'sequential';
```
Prefer enums over raw strings

tcarrio commented on src/implementation/multiprovider/strategy/ComparisonStrategy.php:
```diff
@@ -0,0 +1,185 @@
+ */
+class ComparisonStrategy extends BaseEvaluationStrategy
+{
+    public string $runMode = 'parallel';
```
Prefer enums over raw strings

tcarrio commented on src/implementation/multiprovider/strategy/StrategyEvaluationContext.php:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Context information for the overall evaluation across all providers.
+ */
+class StrategyEvaluationContext
```
It would make sense to move the invariant check for flag type vs default value up into the context or earlier in the evaluation logic

Tmakinde replied:
Hey Tom,

I saw all your comments, and I will make changes to them.

Thank you

tcarrio reviewed (commented):
I left some suggestions for the current PR state.

I am on vacation so can't provide a deeper review for some time.

Tmakinde committed:
chore: Refactor default value check for readability

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: Refactor default value check for readability

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: Use enum instead of string for provider runmode

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: Use enum instead of string for provider runmode

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: Move invariant check for flag type vs default value up into the context

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: Move invariant check for flag type vs default value up into the context

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: Add more test coverage for strategy context and evaluation

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: Add more test coverage for strategy context and evaluation

Signed-off-by: tmakinde <makindet74@gmail.com>

tcarrio commented on src/implementation/multiprovider/Multiprovider.php:
```diff
@@ -0,0 +1,371 @@
+            }
+
+            // Evaluate provider
+            $result = $this->evaluateProvider($provider, $providerName, $baseContext);
```
What is the purpose of creating the `StrategyProviderContext` on 218 if it's not used here?

tcarrio replied:
Nevermind, I can see we have `baseContext` and `perProviderContext`, but they're not actually the same type of class. The naming here can lead to confusion.

Probably better to call these `$evaluationContext` and `$providerContext`.

tcarrio commented on src/implementation/multiprovider/Multiprovider.php:
```diff
@@ -0,0 +1,371 @@
+use function strtolower;
+use function trim;
+
+class Multiprovider extends AbstractProvider
```
This is a very small thing, but the OpenFeature specification calls this `MultiProvider` not `Multiprovider`.


```suggestion
class MultiProvider extends AbstractProvider
```

tcarrio commented on src/implementation/multiprovider/Multiprovider.php:
```diff
@@ -0,0 +1,371 @@
+        'name', 'provider',
+    ];
+
+    public const NAME = 'Multiprovider';
```
```suggestion
    public const NAME = 'MultiProvider';
```

tcarrio commented:
@Tmakinde Beyond the suggestions in the PR thus far, would you be interested in being included as a code owner for the multiprovider namespace?

If so, feel free to update the code owners as part of this pull request.

tcarrio commented on src/implementation/multiprovider/Multiprovider.php:
```diff
@@ -0,0 +1,127 @@
+                }
+                $this->providersByName[$name] = $entry['provider'];
+            }
+        }
```
@Tmakinde just mentioning that this would be also a good idea as the current logic may not capture this duplicate name scenario for non-generated names.

Tmakinde committed:
refactor: improve multiprovider implementation with cleaner abstractions

- Rename StrategyEvaluationContext → StrategyContext for clarity
- Rename StrategyPerProviderContext → ProviderContext for conciseness
- Remove redundant type validation from StrategyContext constructor
  (validation now happens via assertions in evaluateProvider)
- Simplify type resolution using match expression instead of switch
- Add type annotation to $defaultValue for Psalm static analysis
- Simplify provider validation with single-pass loop and better error messages
- Streamline provider registration logic (removed redundant duplicate check)
- Add aggregateErrors() helper in BaseEvaluationStrategy to DRY error collection
- Rename class from Multiprovider → MultiProvider (including filename)

This refactoring reduces code duplication (~60 lines), improves maintainability,
and makes the codebase more idiomatic with modern PHP patterns.

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
test: update tests for refactored multiprovider implementation

- Rename test files: StrategyEvaluationContextTest → StrategyContextTest
- Rename test files: StrategyPerProviderContextTest → ProviderContextTest
- Update test assertions to match new error messages
  - 'Duplicate provider names found' → 'Duplicate provider name'
  - 'Each provider data entry must have...' → 'Provider name cannot be empty'
  - 'Unsupported keys in provider data entry' → 'Unsupported keys: <list>'
- Update MultiProviderStrategyTest to pass ProviderContext to strategy methods
- Remove validation tests from StrategyContextTest (validation removed from constructor)
- All 197 tests pass with 433 assertions

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
chore: add code ownership for multiprovider namespace

Add @tmakinde as code owner for the multiprovider implementation.
This ensures appropriate review oversight for future changes to this
feature area.

Signed-off-by: tmakinde <makindet74@gmail.com>

tcarrio commented on src/implementation/multiprovider/strategy/BaseEvaluationStrategy.php:
```diff
@@ -0,0 +1,108 @@
+            }
+        }
+
+        return $errors !== [] ? $errors : null;
```
I would prefer checking length of the array instead of checking whether it is equal to an empty array.

tcarrio replied:
Similar to elsewhere, use count.

tcarrio commented on src/implementation/multiprovider/MultiProvider.php:
```diff
@@ -0,0 +1,372 @@
+
+    /**
+     * Generate a unique provider name by appending a count suffix if necessary.
+     * E.g., if "ProviderA" is used twice, the second instance becomes "ProviderA_2".
```
The implementation lowercases the name so the example should be updated to follow that expectation.

tcarrio commented on src/implementation/multiprovider/MultiProvider.php:
```diff
@@ -0,0 +1,372 @@
+     */
+    private function generateUniqueName(string $baseName, array &$counts): string
+    {
+        $key = strtolower($baseName);
```
This seems fine, but because you're requiring normalization by lowercasing names, the mapping of providers can cause something like "Provider_A" and "provider_a" to be detected as duplicate, then the second one will be rewritten with the count suffix.

This is likely fine, just making a note. It might be worth mentioning this in the docstring.

Tmakinde replied:
I have made necessary changes for this.

tcarrio commented on tests/unit/MultiProviderStrategyTest.php:
```diff
@@ -0,0 +1,224 @@
+        $strategy = new FirstMatchStrategy();
+
+        $error = new class extends Exception implements ThrowableWithResolutionError {
+            public function getResolutionError(): \OpenFeature\interfaces\provider\ResolutionError
```
Since this is a test file, it's not as big an issue, but prefer to `use` the value namespace and reference by the class name.


```suggestion
            public function getResolutionError(): ResolutionError
```

tcarrio reviewed (approved):
No serious blockers, but some small updates you can make before we wrap. Now approving as the current implementation looks good to me.

Tmakinde committed:
feat(multiprovider): add case-insensitive duplicate provider name validation

- Update validateProviderData to use case-insensitive comparison
- Provider names 'ProviderA' and 'providera' now treated as duplicates
- Add test case to verify case-insensitive duplicate detection
- Original case preserved in error messages

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
refactor(multiprovider): code refinements for clarity and consistency

- Use count() > 0 instead of !== [] in aggregateErrors for explicit intent
- Keep case-sensitive naming in generateUniqueName (ProviderA_2 vs providera_2)
- Add import alias for ResolutionError interface in tests for clarity

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
docs(multiprovider): add comprehensive MultiProvider documentation

- Add MultiProvider to features table with implemented status
- Document MultiProvider usage under Providers section
- Explain three evaluation strategies (FirstMatchStrategy, FirstSuccessfulStrategy, ComparisonStrategy)
- Include provider naming conventions (explicit and auto-generated)
- Add strategy comparison table for easy selection
- Document error handling and aggregation behavior
- Provide complete working examples for each use case

Signed-off-by: tmakinde <makindet74@gmail.com>

jonathannorris commented on src/implementation/multiprovider/MultiProvider.php:
```diff
@@ -0,0 +1,372 @@
+     *
+     * @return array<int, ProviderResolutionResult> Array of resolution results from evaluated providers.
+     */
+    private function evaluateParallel(StrategyContext $strategyContext): array
```
`evaluateParallel` is a plain sequential `foreach` loop — it's identical to `evaluateSequential` except it skips the `shouldEvaluateNextProvider` check. There's no concurrent execution.

I'd recommend renaming the run mode to `EVALUATE_ALL` and the method to `evaluateAll` — that accurately describes what it actually does (evaluate every provider regardless of outcome) without implying concurrency. The `PARALLEL` name comes from the js-sdk reference which uses `Promise.all()`, but that doesn't translate to PHP without an async runtime like Fibers or ReactPHP.

Tmakinde replied:
Got it!

jonathannorris commented on src/implementation/multiprovider/strategy/ComparisonStrategy.php:
```diff
@@ -0,0 +1,180 @@
+     *
+     * @return FinalResult The final result of the evaluation
+     */
+    public function determineFinalResult(
```
Two issues here:

**1. Behavior diverges from the js-sdk reference.** In the [reference implementation](https://github.com/open-feature/js-sdk/blob/main/packages/shared/src/provider/multi-provider/strategies/comparison-strategy.ts), `ComparisonStrategy` stops immediately and returns all errors if *any* provider errors during evaluation. The PHP implementation instead separates errors from successes and returns a result if any provider succeeds. These are different contracts for the same strategy.

Also, the js-sdk requires a `fallbackProvider` as a mandatory constructor argument and throws if it's not found in the results. The PHP version makes it optional and silently falls back to the first result — which makes the fallback semantics much harder to reason about.

**2. The README docs don't match the implementation.** The docs say:
- *"For numbers: Returns the highest value"*
- *"For booleans: `true` wins over `false`"*

But `determineFinalResult` does a strict equality comparison (`$details->getValue() !== $firstValue`) — there's no "highest wins" logic. Either the docs need to be corrected to describe equality comparison, or the aggregation behavior the docs promise needs to be implemented.

jonathannorris commented on src/implementation/multiprovider/MultiProvider.php:
```diff
@@ -0,0 +1,372 @@
+    /**
+     * Evaluate a single provider and return result with error handling.
+     */
+    private function evaluateProvider(ProviderContext $providerContext, StrategyContext $strategyContext): ProviderResolutionResult
```
`evaluateProvider` calls the provider's resolve methods directly, bypassing `$provider->getHooks()` entirely. Any hooks registered on a sub-provider will be silently ignored.

The js-sdk reference runs each sub-provider's `before`, `after`, `error`, and `finally` hooks in sequence around each provider call via a dedicated `HookExecutor`. The PHP `Provider` interface extends `HooksGetter` (Requirement 2.10), so the mechanism is there — it just isn't wired up here.

Happy to track this as a follow-up PR, but I'd recommend calling out the gap explicitly in the docs so users know sub-provider hooks aren't executed yet.

Tmakinde replied:
I have added this to the Readme to be a follow-up implementation

jonathannorris commented on src/implementation/multiprovider/MultiProvider.php:
```diff
@@ -0,0 +1,372 @@
+     * @param mixed $defaultValue The default value to return.
+     * @param array<int, array{providerName: string, error: Throwable}>|null $errors Array of errors encountered during evaluation.
+     */
+    private function createErrorResolution(string $flagKey, mixed $defaultValue, ?array $errors): ResolutionDetails
```
`createErrorResolution` builds a resolution with no value set. The spec (Requirement 4.1.2) requires the default value be returned on error — `$defaultValue` is already in scope, it just needs to be included in the builder call.

Tmakinde replied:
Fixed

jonathannorris commented on src/implementation/multiprovider/ProviderResolutionResult.php:
```diff
@@ -0,0 +1,54 @@
+        return $this->error !== null;
+    }
+
+    public function isSuccessful(): bool
```
`isSuccessful()` only checks whether a `Throwable` was thrown. A provider can also return a `ResolutionDetails` with `getError()` set (a soft error, no exception thrown). In that case `isSuccessful()` returns `true` and the strategy would treat an error-state result as a successful resolution.

The check should also verify that `$this->details->getError() === null`.

Tmakinde replied:
Good catch!

I have fixed it.

Thank you @jonathannorris

jonathannorris commented on src/implementation/multiprovider/MultiProvider.php:
```diff
@@ -0,0 +1,372 @@
+     * @param mixed $defaultValue The default value to return.
+     * @param array<int, array{providerName: string, error: Throwable}>|null $errors Array of errors encountered during evaluation.
+     */
+    private function createErrorResolution(string $flagKey, mixed $defaultValue, ?array $errors): ResolutionDetails
```
When all providers fail, `createErrorResolution` produces a generic `"Multi-provider evaluation failed with 3 provider error(s)"` message and discards the individual errors. The js-sdk throws an `AggregateError` containing each provider's error, which is essential for debugging which provider failed and why.

I'd recommend at minimum concatenating each provider name and error message into the returned error string, since PHP doesn't have a native `AggregateError` equivalent.

Tmakinde replied:
I updated the method to produce more detail error.

jonathannorris commented on tests/unit/MultiproviderTest.php:
```diff
@@ -0,0 +1,248 @@
+        ];
+
+        $multiprovider = new MultiProvider($providerData);
+        $this->assertInstanceOf(Multiprovider::class, $multiprovider);
```
`Multiprovider::class` (lowercase `p`) doesn't match the `use ... MultiProvider` import — PHP `use` aliases are case-sensitive, so this resolves to a nonexistent class. Should be `MultiProvider::class`.

Tmakinde replied:
Fixed.

jonathannorris commented on src/implementation/multiprovider/MultiProvider.php:
```diff
@@ -0,0 +1,372 @@
+     *
+     * @throws InvalidArgumentException If unsupported keys, invalid names, or duplicate names are found.
+     */
+    private function validateProviderData(array $providerData): void
```
`validateProviderData` detects duplicate explicit names case-insensitively (`strtolower`), but `registerProviders` stores and looks up names case-sensitively (`isset($this->providersByName[$name])`). This means the duplicate guard in `registerProviders` could be bypassed by an auto-generated name that differs from an explicit name only in case.

I'd recommend using a consistent casing strategy throughout — either normalize all names to lowercase on storage, or drop the case-insensitive check in validation.

Tmakinde replied:
Hey @jonathannorris 
You're right.

I have made the necessary changes.

Tmakinde committed:
feat(multiprovider): align ComparisonStrategy with JS-SDK spec and improve docs

BREAKING CHANGE: ComparisonStrategy constructor signature changed

- ComparisonStrategy now requires Provider instance as fallback (not optional string name)
- Implements fail-fast error handling - returns immediately if ANY provider errors
- Fallback provider only used for value mismatches (not error recovery)
- Strict equality (===) comparison for all value types

Error Handling:
- Enhanced error aggregation with provider-level details
- Format: "Multi-provider evaluation failed with X provider error(s): [ProviderA]: message"
- Separated aggregation logic into dedicated aggregateErrorMessage() method

Provider Naming:
- Fixed case-sensitivity inconsistency in validation and storage
- All provider names normalized to lowercase for case-insensitive lookup
- Auto-generated names also lowercased

Documentation:
- Restructured README to match JS-SDK clarity and flow
- Added comprehensive Custom Strategies section with working examples
- Documented known limitation: sub-provider hooks not executed
- Updated all examples and use cases
- Added error aggregation examples

Testing:
- Updated ComparisonStrategyTest for new fail-fast behavior
- Added tests for fallback provider requirement
- All 198 tests passing with proper type safety

Renamed RunMode::PARALLEL to RunMode::EVALUATE_ALL for clarity

Signed-off-by: tmakinde <makindet74@gmail.com>

Tmakinde committed:
test(multiprovider): improve ComparisonStrategy test coverage

Add edge case tests:

- Test single provider scenario
- Test mismatch without onMismatch callback
- Test getter methods for fallbackProvider and onMismatch
- Test getOnMismatch returns null when not provided

Signed-off-by: tmakinde <makindet74@gmail.com>

jonathannorris commented on src/implementation/multiprovider/strategy/ComparisonStrategy.php:
```diff
@@ -0,0 +1,174 @@
+            // Check for value mismatch using strict equality
+            $details = $resolution->getDetails();
+            if ($details !== null) {
+                if ($value !== null && $value !== $details->getValue()) {
```
Mismatch detection uses `$value !== null` as a sentinel to know if a value has been seen, but this breaks when the first provider returns `null` — subsequent differing values won't be detected as mismatches. The js-sdk avoids this with `typeof value !== 'undefined'`; the PHP equivalent is an explicit `$valueSet` flag:

```php
$valueSet = false;
$value = null;
// ...
if ($valueSet && $value !== $details->getValue()) {
    $mismatch = true;
} else {
    $value = $details->getValue();
    $valueSet = true;
}
```

This is the last outstanding issue from the review — everything else looks good!

Tmakinde replied:
Good catch!

I have added the suggested changes.

Thank you @jonathannorris

Tmakinde committed:
fix: Fix issue that cab break comparison when value returned as null

Signed-off-by: tmakinde <makindet74@gmail.com>

jonathannorris commented on README.md:
```diff
@@ -164,6 +165,358 @@ $api->setProvider(new MyProvider());
+
+1. **FirstMatchStrategy** (default): Evaluates all providers in order and returns the first successful result. Providers that indicate `FLAG_NOT_FOUND` error will be skipped and the next provider will be evaluated. Any other error will cause the operation to fail and the error to be returned.
+2. **FirstSuccessfulStrategy**: Evaluates all providers in order and returns the first successful result. Any error will cause that provider to be skipped. If no successful result is returned, the set of errors will be returned.
+3. **ComparisonStrategy**: Evaluates all providers in parallel. If every provider returns a successful result with the same value, then that result is returned. Otherwise, an error is returned immediately if any provider errors. When values do not agree, an optional callback will be executed to notify you of the mismatch, and the configured "fallback provider" value will be used. This can be usef ...
```
nit: "Evaluates all providers in parallel" appears here and on line 292, but the implementation is sequential — `evaluateAll` runs providers one at a time. The Behavior section already says it correctly ("Evaluates ALL providers sequentially"). Just needs the intro sentences updated to match.

tcarrio replied:
This is the terminology defined in the OpenFeature JS SDK documentation for the strategy, Especially given there isn't a parallel with the PHP SDK we should lean more into the OpenFeature doc for it instead

https://openfeature.dev/blog/openfeature-multi-provider-release/#comparison-strategy

Tmakinde replied:
Got it!

I have updated the readme

jonathannorris commented on src/implementation/multiprovider/strategy/BaseEvaluationStrategy.php:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return bool True to track with this provider, false to skip
+     */
+    public function shouldTrackWithThisProvider(
```
nit: `shouldTrackWithThisProvider` is correctly defined per Appendix A, but `MultiProvider` never calls it — the PHP `Provider` interface has no `track()` method. Worth adding a brief doc comment so custom strategy implementors know it has no effect yet, e.g. "Not called until `track()` is added to the PHP `Provider` interface."

Tmakinde replied:
I updated the doc block.

jonathannorris reviewed (approved):
See the last two nits, but looks good! Thanks for implementing this!

Tmakinde committed:
docs(multiprovider): clarify evaluation terminology and add track() note

Signed-off-by: tmakinde <makindet74@gmail.com>
